libs/gltfio/src/CPUMorpher.cpp

```diff
@@ -0,0 +1,215 @@
+            // handled correctly.
+            //
+            // filament shaders deal with tangent frame quaternions instead of normal vectors.
+            // But in case of missing inputs, we get a invalid quaternion (0, 0, 0, 0) instead of
+            // a identity quaterion. This leads to the normals being morphed even no inputs are
+            // given.
+            // 
+            // To fix this, we put a empty morph target at slot 2 and give it a weight of -1.
+            // This won't affect the vertex positions but will cancel out the normal values.
+            //
+            // Note that for this to work, at least two morph targets are required.
+            float* data2 = (float*) malloc(size);
+            memset(data2, 0, size);
+            VertexBuffer::BufferDescriptor bd2(data2, size, FREE_CALLBACK);
+            prim.morphBuffer2 = BufferObject::Builder().size(size).build(engine);
+            prim.morphBuffer2->setBuffer(engine, std::move(bd2));
+            vb->setBufferObjectAt(engine, prim.baseSlot+1, prim.morphBuffer2);
+        }
+        VertexBuffer::BufferDescriptor bd(data, size, FREE_CALLBACK);
+        prim.morphBuffer1->setBuffer(engine, std::move(bd));
+        vb->setBufferObjectAt(engine, prim.baseSlot, prim.morphBuffer1);
+    }
+    renderableManager->setMorphWeights(renderable, {1, -1, 0, 0});
+}
+
+void CPUMorpher::addPrimitive(cgltf_mesh const* mesh, int primitiveIndex, TableEntry* entry) {
+    auto& engine = *mAsset->mEngine;
+    const cgltf_primitive& cgltf_prim = mesh->primitives[primitiveIndex];
+
+    int posIndex = findPositionAttribute(cgltf_prim);
+    if (posIndex < 0) return;
+
+    VertexBuffer* vertexBuffer = mAsset->mMeshCache.at(mesh)[primitiveIndex].vertices;
+    int slot = determineBaseSlot(cgltf_prim);
+    entry->primitives.push_back({ vertexBuffer, slot });
+
+    auto& primitive = entry->primitives.back();
+
+    cgltf_attribute& positionAttribute = cgltf_prim.attributes[posIndex];
+    size_t dim = cgltf_num_components(positionAttribute.data->type);
+    primitive.floatsCount = positionAttribute.data->count * dim;
+
+    std::vector<GltfTarget>& targets = primitive.targets;
+
+    for (int targetIndex = 0; targetIndex < cgltf_prim.targets_count; targetIndex++) {
+        const cgltf_morph_target& morphTarget = cgltf_prim.targets[targetIndex];
+        for (cgltf_size aindex = 0; aindex < morphTarget.attributes_count; aindex++) {
+            const cgltf_attribute& attribute = morphTarget.attributes[aindex];
+            const cgltf_accessor* accessor = attribute.data;
+            const cgltf_attribute_type atype = attribute.type;
+
+            // only works for morphing of positions for now
+            if (atype == cgltf_attribute_type_position) {
+                targets.push_back({targetIndex, atype, accessor->type});
+
+                // expect the morph target to use a sparse accessor
```

Comment:
These are very specific limitations for now. We need to be careful if we change how thunder/assimp export models.